File: app/src/main/java/com/juuda/droidmock/alarm/AlarmPlugin.java

package com.juuda.droidmock.alarm;

import android.app.AlarmManager;
import android.app.Application;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.util.Log;

import com.facebook.stetho.dumpapp.DumpException;
import com.facebook.stetho.dumpapp.DumperContext;
import com.facebook.stetho.dumpapp.DumperPlugin;
import com.juuda.droidmock.CommonActivity;
import com.juuda.droidmock.mock.MockUtils;
import com.juuda.droidmock.stetho.StethoHelper;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import java.io.PrintWriter;


public class AlarmPlugin implements DumperPlugin{
    private static final String TAG = "AlarmPlugin";
    private  Application mApplication;
    private BroadcastReceiver mReceiver;
    private PrintWriter mPrintWriter;

    private final String ALARM_ACTION = "com.android.deskclock.ALARM_ALERT";
    public AlarmPlugin(Application application) {
        mApplication = application;
    }

    public void cancelTask() {
        AlarmManager am = (AlarmManager)mApplication.getSystemService(Context.ALARM_SERVICE);
        PendingIntent operation = PendingIntent.getBroadcast(mApplication, 0, new Intent(ALARM_ACTION),
                PendingIntent.FLAG_UPDATE_CURRENT);
        operation.cancel();//works
        am.cancel(operation);//works
    }

    public void start(int delay) {
        mReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                Log.e(TAG, "received.. TAG");
                mPrintWriter.println("alarm received" + intent.getAction());
                mPrintWriter.close();
            }
        };
        mApplication.getApplicationContext().registerReceiver(mReceiver, new IntentFilter(ALARM_ACTION));
        AlarmManager am = (AlarmManager)mApplication.getSystemService(Context.ALARM_SERVICE);
        PendingIntent operation = PendingIntent.getBroadcast(mApplication, 0,
                new Intent(ALARM_ACTION),
                PendingIntent.FLAG_UPDATE_CURRENT);

        long currentMills = System.currentTimeMillis();
        am.set(AlarmManager.RTC_WAKEUP , currentMills + delay, operation);
    }

    @Override
    public String getName() {
        return "alarm";
    }

    private Options getOptions() {
        return new Options()
                .addOption("t", "type", true, "commnd type")
                .addOption("d", "delay", true, "delay");
    }

    @Override
    public void dump(DumperContext dumpContext) throws DumpException {
        mPrintWriter =  StethoHelper.getPrintWriter(dumpContext);
        try {
            CommandLine commandLine = StethoHelper.parse(dumpContext, getOptions());
            String type = commandLine.getOptionValue('t', "start");
            int delay = Integer.parseInt(commandLine.getOptionValue('d'));
            if ("start".equals(type)) {
                start(delay);
            } else if ("cancel".equals(type)) {
                cancelTask();
            }
        } catch (ParseException e) {
            mPrintWriter.println(Log.getStackTraceString(e));
        } finally {
            mPrintWriter.flush();
        }
    }

}

File: app/src/main/java/com/juude/droidmocks/alarm/AlarmPlugin.kt
package com.juude.droidmocks.alarm

import android.app.AlarmManager
import android.app.Application
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.util.Log

import com.facebook.stetho.dumpapp.DumpException
import com.facebook.stetho.dumpapp.DumperContext
import com.facebook.stetho.dumpapp.DumperPlugin
import com.juude.droidmocks.stetho.StethoHelper

import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Options
import org.apache.commons.cli.ParseException

import java.io.PrintWriter


class AlarmPlugin(private val mApplication: Application) : DumperPlugin {
    private var mReceiver: BroadcastReceiver? = null
    private var mPrintWriter: PrintWriter? = null

    private val ALARM_ACTION = "com.android.deskclock.ALARM_ALERT"

    fun cancelTask() {
        val am = mApplication.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val operation = PendingIntent.getBroadcast(mApplication, 0, Intent(ALARM_ACTION),
                PendingIntent.FLAG_UPDATE_CURRENT)
        operation.cancel()//works
        am.cancel(operation)//works
    }

    fun start(delay: Int) {
        mReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                Log.e(TAG, "received.. TAG")
                mPrintWriter!!.println("alarm received" + intent.action)
                mPrintWriter!!.close()
            }
        }
        mApplication.applicationContext.registerReceiver(mReceiver, IntentFilter(ALARM_ACTION))
        val am = mApplication.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val operation = PendingIntent.getBroadcast(mApplication, 0,
                Intent(ALARM_ACTION),
                PendingIntent.FLAG_UPDATE_CURRENT)

        val currentMills = System.currentTimeMillis()
        am.set(AlarmManager.RTC_WAKEUP, currentMills + delay, operation)
    }

    override fun getName(): String {
        return "alarm"
    }

    private val options: Options
        get() = Options().addOption("t", "type", true, "commnd type").addOption("d", "delay", true, "delay")

    override fun dump(dumpContext: DumperContext) {
        mPrintWriter = StethoHelper.getPrintWriter(dumpContext)
        try {
            val commandLine = StethoHelper.parse(dumpContext, options)
            val type = commandLine.getOptionValue('t', "start")
            val delay = Integer.parseInt(commandLine.getOptionValue('d'))
            if ("start" == type) {
                start(delay)
            } else if ("cancel" == type) {
                cancelTask()
            }
        } catch (e: ParseException) {
            mPrintWriter!!.println(Log.getStackTraceString(e))
        } finally {
            mPrintWriter!!.flush()
        }
    }

    companion object {
        private val TAG = "AlarmPlugin"
    }

}